### PR TITLE
Post-merge cleanup: tests, fixtures, dist freshness check, publish=false fix, rust-version/edition outputs, 1.0.0

### PR DESCRIPTION
**Pull request by SVilgelm**
Follow-up to #19, addressing every item from the review and bumping to **1.0.0**.

### Code

- **Bug fix: `publish = false` was incorrectly classified as publishable.** Cargo represents `publish = false` as `[]` in metadata, not `false`. The previous check `pkg.publish !== false` always returned true. Now: `null` and non-empty arrays are publishable; an empty array is excluded.
- **Default empty arrays for `packages`/`publish`/`matrix`.** Previously, if `cargo metadata` ever returned without a `packages` key, downstream `fromJson(...outputs.matrix)` would crash on an empty string.
- **Dropped `@actions/github`.** It was a runtime dep but never imported; bundling its `@octokit/*` graph was bloating `dist/` for nothing.
- **Refactor: split into `lib.js` (testable) + one-line `index.js` entry.** ncc still builds `index.js` → `dist/index.js`; tests import from `lib.js` directly.
- **Dropped dead `hasOwnProperty` guards** (cargo metadata always emits both `publish` and `features`).

### New outputs

- **`rust-version`** — workspace MSRV: the highest `rust-version` declared by any package, compared numerically (so `1.10` beats `1.9`). Empty string when no package declares one. Direct use case: feed into `actions-rust-lang/setup-rust-toolchain`.
- **`edition`** — the highest edition used by any package (e.g. `2021`, `2024`). Empty string for an empty workspace.

Both are raw scalar strings (not JSON-encoded), so consumers can read them directly with `${{ steps.x.outputs.rust-version }}` — no `fromJson` needed.

### Tests

- **Unit tests** (`tests/parse-metadata.test.js`) using `node:test` — no extra deps. 14 tests covering empty workspace, missing `packages` key, matrix-vs-features cases, every `publish` shape, MSRV max-across-packages with the `1.10 > 1.9` numeric edge case, and edition max-across-packages.
- **In-tree fixture projects** under `tests/fixtures/` exercise every code path end-to-end without depending on external repos:
  - `single-public/` — one crate, default-publishable, `rust-version = "1.85"`, `edition = "2021"`.
  - `single-private/` — one crate with `publish = false`, no `rust-version` (regression test for the publish-false bug + MSRV-empty case).
  - `workspace-mixed/` — workspace with three members: `app` (1.85 / 2021), `private-lib` (no MSRV, `publish = false`), and `feature-lib` (1.90 / 2024 with default/foo/bar features). Exercises workspace iteration, the publish filter, per-feature matrix expansion, and the MSRV/edition max-across-packages logic together.
  - `toolchain-pinned/` — one crate with `rust-toolchain.toml` pinning `1.85.0`. Replaces the previous inline `ToolchainPin` job that scaffolded a project at runtime.

### CI

- **`Check` job** now also runs `npm test`.
- **`Dist` job** rebuilds and asserts `git diff --exit-code dist/`, so PRs with stale bundled output are blocked instead of being patched after-the-fact.
- **`Fixtures` job** runs the action against each in-tree fixture and asserts exact outputs (now including `rust-version` and `edition`):
  - Literal string equality on simple fixtures.
  - `jq`-based sorted set comparison for `workspace-mixed` (stable across cargo versions, since feature ordering is alphabetical).
  - For `toolchain-pinned`, also greps `rustup toolchain list` to confirm the install actually ran end-to-end — a future regression that silently skipped `ensureToolchain` would fail this check.
- **`npm install` → `npm ci`** in both workflows for deterministic installs.
- **`actions/setup-node`** added to both jobs for cached Node + npm installs (pinned by SHA, matching the existing convention).
- **`update-dist.yml`** now uses a per-run branch (`update-dist-<run_id>`) and opens a PR via `gh pr create`, instead of failing on the second run when the fixed branch already exists.

### Docs

- README and action.yml document all outputs (including the two new ones), the toolchain auto-install, the matrix-features behavior, and the `publish = false` exclusion.

### Misc

- LICENSE copyright fixed to `Sergey Vilgelm` (was `SV Tools`).
- **Bumped to `1.0.0`** — API (inputs/outputs) is stable, the action has been in active use, and the cleanup in this PR + #19 closes the major outstanding behavioral bugs.

### Breaking change to flag for the v1.0 cut

Packages with `publish = false` are now (correctly) excluded from the `publish` output. Prior versions included them due to the `pkg.publish !== false` bug. Any consumer that relied on the bug will see a behavior change.

## Test plan

- [x] `npm run all` — fmt + build + test all green locally; 14/14 unit tests pass.
- [x] Bundled `dist/index.js` smoke-tested against all four fixtures locally; outputs match expectations including `rust-version` and `edition`.
- [ ] CI: `Check`, `Dist`, `Test`, `Fixtures`, `Roas`, `Nuspec` should all pass.
- [ ] After merge, tag `v1.0.0` and move/refresh the floating `v1` tag.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,11 +9,13 @@
     "Metadata"
   ],
   "main": "index.js",
+  "type": "module",
   "scripts": {
-    "all": "npm run fmt && npm run build",
+    "all": "npm run fmt && npm run build && npm test",
     "build": "ncc build index.js --license licenses.txt",
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply package.json:15]: Verified working — ncc 0.38.4 detects ESM input (the `import` statements + `"type": "module"`) and bundles to ESM, then writes its own `dist/package.json` with `{"type":"module"}` to keep the bundled entry consistent. Local smoke test of `node dist/index.js` against all four fixtures produces correct outputs, and the new `Dist` CI job rebuilds + asserts no diff so any future regression here would be caught.

File: .github/workflows/update-dist.yml [hunk 2]
```diff
@@ -15,21 +15,32 @@ jobs:
...
+      pull-requests: write
     steps:
       - uses: actions/checkout@1af3b93b6815bc44a9784bd300feb67ff0d1eeb3 # v5.5.0
-      - run: npm install
+      - uses: actions/setup-node@49933ea5288caeca8642d1e84afbd3f7d6820020 # v4.4.0
+        with:
+          node-version-file: .nvmrc
+          cache: npm
+      - run: npm ci
       - run: npm run all
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply .github/workflows/update-dist.yml]: Good catch — fixed in 1ede0b0. Changed `npm run all` → `npm run build` and `git add .` → `git add dist/` so the bot only commits the bundled output, not unrelated formatting/test-induced drift.

File: .github/workflows/update-dist.yml [hunk 3]
```diff
@@ -15,21 +15,32 @@ jobs:
...
-            git push origin update-dist
+          if [ -z "$(git status --short)" ]; then
+            echo "dist/ already up to date"
+            exit 0
           fi
+          git config --global user.name 'SV Tools Bot'
+          git config --global user.email 'sv.go.tools@gmail.com'
+          BRANCH="update-dist-${{ github.run_id }}"
+          git checkout -b "$BRANCH"
+          git add .
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply .github/workflows/update-dist.yml]: Same issue as above — addressed in 1ede0b0 by switching `npm run all` → `npm run build` and `git add .` → `git add dist/`.

File: lib.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+      publish.push(pkg.name);
+    }
+    const features = Object.keys(pkg.features ?? {});
+    if (features.length === 0) {
+      matrix.push(`--package=${pkg.name}`);
+    } else {
+      for (const feature of features) {
+        matrix.push(`--package=${pkg.name} --features=${feature}`);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply lib.js]: Fixed in 1ede0b0 — `parseMetadata` now sorts features explicitly. Cargo currently emits features alphabetically (BTreeMap) so it happened to be stable already, but the explicit sort makes the order contractual on our end. Added a unit test using reverse-alphabetical insertion order that would fail if the sort step were ever dropped.

File: lib.js [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+
+export function runCargoMetadata(manifestPath) {
+  return new Promise((resolve, reject) => {
+    const cmd = spawn("cargo", [
+      "metadata",
+      "--manifest-path=" + manifestPath,
+      "--no-deps",
+      "--format-version",
+      "1",
+    ]);
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply lib.js]: Real bug — fixed in 1ede0b0. `runCargoMetadata` now spawns with `cwd = dirname(resolvePath(manifestPath))`, matching `ensureToolchain`. Without this, `ensureToolchain` would install the pinned toolchain in the right directory, but `cargo metadata` running from `process.cwd()` would cause rustup to look for `rust-toolchain.toml` in the wrong place (rustup walks up, not down) and fall back to the default toolchain — so the install and the actual invocation could silently use different toolchains. Also passes the absolute manifest path now, so a relative input does not get re-resolved against the new cwd.

File: lib.js [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+  setOutput("rust-version", rustVersion);
+  setOutput("edition", edition);
+}
+
+export async function run() {
+  const manifestPath = getInput("manifest-path", { required: true });
+  ensureToolchain(manifestPath);
+  const metadata = await runCargoMetadata(manifestPath);
+  writeOutputs(metadata);
+}
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply lib.js:171]: Fixed in 1ede0b0 — dropped `{ required: true }`. The default in `action.yml` makes the input always present, so the runtime check was redundant and inconsistent with the published `required: false` contract.

File: tests/fixtures/workspace-mixed/feature-lib/Cargo.toml [hunk 7]
```diff
@@ -0,0 +1,13 @@
+[package]
+name = "feature-lib"
+version = "0.1.0"
+edition = "2024"
+rust-version = "1.90"
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply tests/fixtures/workspace-mixed/feature-lib/Cargo.toml:5]: Tradeoff acknowledged, leaving as-is. Pinning a toolchain via `rust-toolchain.toml` would change what this fixture tests — `workspace-mixed` is specifically the no-toolchain-file code path, complementary to `toolchain-pinned`. ubuntu-latest`s preinstalled stable Rust is regularly refreshed and supports edition 2024 (stabilized in 1.85.0). If we ever hit flakiness here, easy fix is either adding a workspace-level rust-toolchain.toml or downshifting feature-lib to edition 2021.

File: .github/workflows/update-dist.yml [hunk 8]
```diff
@@ -15,21 +15,32 @@ jobs:
...
+          git commit -m "Update Dist"
+          git push origin "$BRANCH"
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply .github/workflows/update-dist.yml]: Same fix as the earlier two on this file — addressed in 1ede0b0 (`npm run all` → `npm run build`, `git add .` → `git add dist/`).

File: package.json [hunk 9]
```diff
@@ -9,11 +9,13 @@
     "Metadata"
   ],
   "main": "index.js",
+  "type": "module",
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply package.json:12]: Already addressed in the previous round of comments — `dist/index.js` is genuinely ESM. ncc 0.38.4 detects the ESM input (the `import` statements + the package-level `"type": "module"`) and bundles to ESM (build log: `Compiling file index.js into ESM`), then writes its own `dist/package.json` with `{"type":"module"}` so the bundled entry stays ESM regardless of the parent package settings. Local smoke test of `node dist/index.js` against all four fixtures produces correct outputs, and the `Dist` CI job rebuilds + asserts no diff so any future regression here would be caught.

File: tests/fixtures/single-public/Cargo.toml [hunk 10]
```diff
@@ -0,0 +1,10 @@
+[package]
+name = "single-public"
+version = "0.1.0"
+edition = "2021"
+rust-version = "1.85"
+publish = true
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply tests/fixtures/single-public/Cargo.toml]: Verified: `cargo metadata` accepts `publish = true` and emits it as `publish: null` (i.e. same as omitting) — no parse failure. That said, `publish = true` is not in the documented Cargo.toml schema (only `false` and a registry-list array are), so I removed it in 9f68ae1 to keep the fixture idiomatic. The fixture name `single-public` already conveys the publishable-by-default intent.

File: lib.js [hunk 11]
```diff
@@ -0,0 +1,171 @@
...
+    rustVersion: rustVersion ?? "",
+    edition: edition ?? "",
+  };
+}
+
+export function writeOutputs(metadata) {
+  const { packages, publish, matrix, rustVersion, edition } =
+    parseMetadata(metadata);
+  setOutput("metadata", JSON.stringify(metadata));
+  setOutput("packages", JSON.stringify(packages));
```
[the remark this answers is not in the document]
Comment (SVilgelm) [reply lib.js:159]: Considered, leaving as-is. A correct skip-detection has to walk up the directory tree like rustup itself does (a `rust-toolchain.toml` may live in a parent of the manifest directory — common in workspaces where the manifest is in `crates/foo/Cargo.toml` and the pin is at the workspace root). Re-implementing that walk just to save the ~200ms `rustup toolchain install --no-self-update` no-op when the default toolchain is already installed (which it always is on GitHub-hosted runners) does not feel like a worthwhile complexity tradeoff. If the latency ever becomes a real concern in profiling we can revisit, but for now keeping the simple unconditional invocation that exactly mirrors what `cargo` itself does on every call.
